### ISSUE TITLE
Zero division in SoftmaxLossLayer when a minibatch contains only ignored labels

### ISSUE THREAD

**Issue by muupan** [BVLC/caffe#2335]
I've noticed that SoftmaxLossLayer divides ```loss``` by zero when

-  ```label``` contains only labels specified as ```ignore_label``` and
- ```normalize``` option set ```true``` (it's the default value),

which would result in loss nan.

https://github.com/BVLC/caffe/blob/c6414ea7cab5917c904729c87f786e0a5909475c/src/caffe/layers/softmax_loss_layer.cpp#L77
https://github.com/BVLC/caffe/blob/c6414ea7cab5917c904729c87f786e0a5909475c/src/caffe/layers/softmax_loss_layer.cpp#L116
https://github.com/BVLC/caffe/blob/c6414ea7cab5917c904729c87f786e0a5909475c/src/caffe/layers/softmax_loss_layer.cu#L56
https://github.com/BVLC/caffe/blob/c6414ea7cab5917c904729c87f786e0a5909475c/src/caffe/layers/softmax_loss_layer.cu#L116

**Comment by weiliu89** [comment]
@muupan That's true. You have to add a if (count == 0) condition, then loss = 0, or don't scale gradient in the backward operation.

**Comment by longjon** [comment]
We should probably at least add a (debug?) check for this condition; you're welcome to PR that.

**Comment by muupan** [comment]
It seems to me that adding ```CHECK``` would be better than ```DCHECK``` for this issue because:
1. This problem can arise only randomly (sometimes only rarely) if data are shuffled, thus it can be very hard to be reproduced.
2. When the problem arises, what one can see is that everything got ```nan``` and there would be few clues for solving the problem.

I can't tell which is a better solution: adding ```CHECK``` or making ```loss = 0```. ```AccuracyLayer``` also has the same issue, so it should be dealt with in a consistent way.

**Comment by shelhamer** [comment]
> I can't tell which is a better solution: adding CHECK or making loss = 0

Better to just handle this and make `loss = 0`, no? It might incidentally happen that an ignore batch is loaded, and this seems like the responsibility of the experimenter more than the framework to check.
